### ISSUE TITLE
Should not use ANSI Windows APIs

### ISSUE THREAD

**Issue by aursulis** [osquery/osquery#2569]
Classes like `PlatformFile` (and numerous others) use ANSI versions of Windows APIs.
An example from https://github.com/facebook/osquery/blob/master/osquery/filesystem/windows/fileops.cpp#L389:

``` c++
  handle_ = ::CreateFileA(path.c_str(),
                          access_mask,
                          0,
                          security_attrs.get(),
                          creation_disposition,
                          flags_and_attrs,
                          nullptr);
```

This will mean that they don't support filenames with Unicode characters that are not in the range of the current ANSI code page. Similar implications for other APIs. Files enumerated through `FindFirstFileA` will have such characters replaced with a question mark, etc.

Basically, no new Windows program should be using the A APIs. All they do anyway is convert their arguments to UTF-16 wide types (according to the current code page) and call the W version of the API.

**Comment by PoppySeedPlehzr** [comment]
Hey @aursulis, thanks for the heads up!  Do you have any good recommendations for bouncing between UTF-8 and UTF-16 data then? A lot of the data we store or pass back is UTF-8 text, thus looking for the Ansi versions of the API function calls seemed like a good idea at the time.

We have a few small helper functions to make the conversions happen, which should allow us to leverage the wide API calls, however we'll hit the same problem in converting things. Do you have any recommendations or guides for the most appropriate way to handle data when we need things to be back in UTF-8 for later purposes?

**Comment by theopolis** [comment]
We could investigate storing data as UTF16 all the way up until logging. At that point we convert the responses into JSON* and write to files/send to TLS endpoints/etc. We have a JSON escape method and we can do the conversion there.

* We use JSON internally for intermediate storage in RocksDB, but that's opaque to the user. It will perform the same JSON escape/unescape during data serialization, but should otherwise not be noticed.

**Comment by aursulis** [comment]
I had a quick look at what APIs are used and in the majority of cases the string you're passing in/receiving out is a path (or a conceptually similar thing, like an environment variable name). So since you use `boost::filesystem` already, why not go for `boost::filesystem::path` or `boost::filesystem::path::string_type` and then convert when necessary (maybe even using the boost `path_traits` converters if one can figure out how they work)

**Comment by scubadriver** [comment]
Please do not use boost.
    DWORD dwFlags = WC_ERR_INVALID_CHARS;//or 0;
        // Convert the wide string to UTF8:
        nResSize = WideCharToMultiByte(CP_UTF8,         // Actual conversion
                                        dwFlags,        // flags
                                        wsz,            // wide string
                                        cchWideChar,    // length of input string. see comments above
                                        sz,             // Output buffer
                                        nBuffSize,      // Size of the buffer available for conversion
                                        NULL,           // default character: must be NULL for Utf8
                                        NULL);          // BOOL to record if default char applied: must be NULL for Utf8
        if (nResSize > 0)  sRet = sz;
Similarly to convert UTF8 to UTF16:
        nRetSize = MultiByteToWideChar(CP_UTF8, nFlags, sz, cch, wsz, nBuffSize);
There are plenty of samples online for this.
Best
